up-streamer/src/ustreamer.rs

```diff
@@ -0,0 +1,986 @@
+
+impl TransportForwarder {
+    async fn new(
+        out_transport: Arc<Mutex<Box<dyn UTransport>>>,
+        message_receiver: Receiver<Arc<UMessage>>,
+    ) -> Self {
+        let out_transport_clone = out_transport.clone();
+        let message_receiver_clone = message_receiver.clone();
+        task::spawn(Self::message_forwarding_loop(
+            UUIDBuilder::build().to_hyphenated_string(),
+            out_transport_clone,
+            message_receiver_clone,
+        ));
+
+        Self {}
+    }
+
+    async fn message_forwarding_loop(
+        id: String,
+        out_transport: Arc<Mutex<Box<dyn UTransport>>>,
+        message_receiver: Receiver<Arc<UMessage>>,
+    ) {
+        while let Ok(msg) = message_receiver.recv().await {
+            debug!(
+                "{}:{}:{} Attempting send of message: {:?}",
+                id,
+                TRANSPORT_FORWARDER_TAG,
+                TRANSPORT_FORWARDER_FN_MESSAGE_FORWARDING_LOOP_TAG,
+                msg
+            );
+
+            let out_transport = out_transport.lock().await;
+            let send_res = out_transport.send(msg.deref().clone()).await;
+            if let Err(err) = send_res {
+                error!(
+                    "{}:{}:{} Sending on out_transport failed: {:?}",
+                    id,
+                    TRANSPORT_FORWARDER_TAG,
+                    TRANSPORT_FORWARDER_FN_MESSAGE_FORWARDING_LOOP_TAG,
+                    err
+                );
+            } else {
+                debug!(
+                    "{}:{}:{} Sending on out_transport succeeded",
+                    id, TRANSPORT_FORWARDER_TAG, TRANSPORT_FORWARDER_FN_MESSAGE_FORWARDING_LOOP_TAG
+                );
+            }
+        }
+    }
+}
+
+const FORWARDING_LISTENER_TAG: &str = "ForwardingListener:";
+const FORWARDING_LISTENER_FN_ON_RECEIVE_TAG: &str = "on_receive():";
+const FORWARDING_LISTENER_FN_ON_ERROR_TAG: &str = "on_error():";
+
+#[derive(Clone)]
+pub(crate) struct ForwardingListener {
+    forwarding_id: String,
+    sender: Sender<Arc<UMessage>>,
+}
+
+impl ForwardingListener {
+    pub(crate) async fn new(forwarding_id: &str, sender: Sender<Arc<UMessage>>) -> Self {
+        Self {
+            forwarding_id: forwarding_id.to_string(),
+            sender,
+        }
+    }
+}
+
+#[async_trait]
+impl UListener for ForwardingListener {
+    async fn on_receive(&self, msg: UMessage) {
+        debug!(
+            "{}:{}:{} Received message: {:?}",
+            self.forwarding_id,
+            FORWARDING_LISTENER_TAG,
+            FORWARDING_LISTENER_FN_ON_RECEIVE_TAG,
+            &msg
+        );
+        if let Err(e) = self.sender.send(Arc::new(msg)).await {
+            error!(
+                "{}:{}:{} Unable to send message to worker pool: {e:?}",
+                self.forwarding_id, FORWARDING_LISTENER_TAG, FORWARDING_LISTENER_FN_ON_RECEIVE_TAG,
+            );
+        }
+    }
+
+    async fn on_error(&self, err: UStatus) {
```

Comment:
I realize now that this function is not helpful without having any *context* in which the error occurred. IMHO we should also pass in the UAttributes?

Reply:
Yeah I think that could be nice. Otherwise you're at the mercy of whatever the `UTransport` impl decided to stick into the message portion of the `UStatus`.

You wanna take the ball and open issue + PR?

Reply:
see https://github.com/eclipse-uprotocol/up-rust/issues/87